source/common/common/trie_lookup_table.h

```diff
@@ -0,0 +1,167 @@
+      std::vector<int32_t> new_children;
+      new_children.reserve(node.min_child_key_ - char_key + node.children_.size());
+      new_children.resize(node.min_child_key_ - char_key, -1);
+      std::move(node.children_.begin(), node.children_.end(), std::back_inserter(new_children));
+      new_children[0] = child_index;
+      node.min_child_key_ = char_key;
+      node.children_ = std::move(new_children);
+      return;
+    }
+    if (char_key >= (node.min_child_key_ + node.children_.size())) {
+      // Expand the vector forwards.
+      node.children_.resize(char_key - node.min_child_key_ + 1, -1);
+      // Fall through to "insert" behavior.
+    }
+    node.children_[char_key - node.min_child_key_] = child_index;
+  }
+
+public:
+  /**
+   * Adds an entry to the Trie at the given Key.
+   * @param key the key used to add the entry.
+   * @param value the value to be associated with the key.
+   * @param overwrite_existing will overwrite the value when the value for a given key already
+   * exists.
+   * @return false when a value already exists for the given key.
+   */
+  bool add(absl::string_view key, Value value, bool overwrite_existing = true) {
+    int32_t current = 0;
+    for (uint8_t c : key) {
+      int32_t next = getChildIndex(current, c);
+      if (next == -1) {
+        next = nodes_.size();
+        nodes_.emplace_back();
+        setChildIndex(current, c, next);
+      }
+      current = next;
+    }
+    if (nodes_[current].value_ && !overwrite_existing) {
+      return false;
+    }
+    nodes_[current].value_ = std::move(value);
+    return true;
+  }
+
+  /**
+   * Finds the entry associated with the key.
+   * @param key the key used to find.
+   * @return the Value associated with the key, or an empty-initialized Value
+   *         if there is no matching key.
+   */
+  Value find(absl::string_view key) const {
+    int32_t current = 0;
+    for (uint8_t c : key) {
+      current = getChildIndex(current, c);
+      if (current == -1) {
+        return {};
+      }
+    }
+    return nodes_[current].value_;
+  }
+
+  /**
+   * Finds the entry with the longest key that is a prefix of the specified key.
+   * Complexity is O(min(longest key prefix, key length)).
+   * @param key the key used to find.
+   * @return a value whose key is a prefix of the specified key. If there are
+   *         multiple such values, the one with the longest key. If there are
+   *         no keys that are a prefix of the input key, an empty-initialized Value.
+   */
+  Value findLongestPrefix(absl::string_view key) const {
+    int32_t current = 0;
+    int32_t result = 0;
+
+    for (uint8_t c : key) {
+      current = getChildIndex(current, c);
+
+      if (current == -1) {
+        return nodes_[result].value_;
+      } else if (nodes_[current].value_) {
+        result = current;
+      }
+    }
+    return nodes_[result].value_;
+  }
+
+private:
+  std::vector<TrieNode> nodes_ = {TrieNode()};
```

Comment:
why is the initializer needed?

Reply:
Added an explanatory comment (it's the root node).